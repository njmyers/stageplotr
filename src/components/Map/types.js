import type { Kit } from '../Kit/types';

export type Map = {
  name: string,
  width: number,
  height: number,
  length: number,
  items: Array<Kit>,
};
