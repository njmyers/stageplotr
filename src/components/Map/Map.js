// @flow
import * as React from 'react';
import uuid from 'uuid/v1';

import Kit from '../Kit';
import Resize from '../Resize';
import type { State as MapState } from './map-reducer';

type State = {};

type Props = {
  map: MapState,
};

class Map extends React.Component<Props, State> {
  svg = null;

  onCreate = () => {
    this.props.createKit();
  };

  render() {
    return (
      <section>
        <svg
          ref={(svg) => (this.svg = svg)}
          width={this.props.map.width}
          height={this.props.map.height}
        >
          {Object.keys(this.props.map.kits).map((id) => {
            return (
              <Resize key={id}>
                <Kit id={id} />
              </Resize>
            );
          })}
        </svg>
        <button onClick={this.onCreate}>Create Kit!</button>
      </section>
    );
  }
}

export default Map;
