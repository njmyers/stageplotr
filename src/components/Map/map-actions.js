const createKit = (id, typeID) => ({
  type: '@MAP/CREATE_KIT',
  id,
  typeID,
});

/**
 * Delete kit object by property id
 * This id functions as a hash table identifier
 * @param  {[type]} value [description]
 * @return {[type]}       [description]
 */
const deleteKit = (id: string) => ({
  type: '@MAP/DELETE_KIT',
  id,
});

const renameMap = (value: string) => ({
  type: '@MAP/RENAME',
  value,
});

const resizeMapWidth = (value: number) => ({
  type: '@MAP/RESIZE_WIDTH',
  value,
});

const resizeMapHeight = (value: number) => ({
  type: '@MAP/RESIZE_HEIGHT',
  value,
});

type Fn<T> = (...args: Array<any>) => T;
type ExtractReturn = <T>(Fn<T>) => T;

export type Action =
  | $Call<ExtractReturn, typeof createKit>
  | $Call<ExtractReturn, typeof deleteKit>
  | $Call<ExtractReturn, typeof renameMap>
  | $Call<ExtractReturn, typeof resizeMapWidth>
  | $Call<ExtractReturn, typeof resizeMapHeight>;

export { createKit, deleteKit, renameMap, resizeMapWidth, resizeMapHeight };
