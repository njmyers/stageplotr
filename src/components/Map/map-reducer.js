// @flow
import kitReducer from '../Kit/kit-reducer';
import type { Kit } from '../Kit/kit-reducer';
import type { Action } from './map-actions';

export type Map = {
  name: string,
  width: number,
  height: number,
  length: number,
  /** Jeff you were right the collection of kits needs to be something
   * besides a simple array
   * This currently is an implementation kind of like a map hash table.
   * This is flow syntax to designate that behavior
   * unknown property key but specific object type
   * https://flow.org/en/docs/types/objects/#toc-objects-as-maps
   */
  kits: {
    [key: string]: Kit,
    __proto__: null,
  },
};

const initialState = {
  name: 'StagePlot',
  width: 100,
  height: 100,
  length: 0,
  // https://davidwalsh.name/object-create-null
  // blank dictionary
  kits: Object.create(null),
};

function mapReducer(state: Map = initialState, action: Action) {
  switch (action.type) {
    // these are integrated with kits
    case '@MAP/CREATE_KIT':
      return {
        ...state,
        length: state.length + 1,
        kits: {
          ...state.kits,
          [action.id]: kitReducer({}, action),
        },
      };
    case '@MAP/DELETE_KIT': {
      // functional object deleting
      const { [action.id]: deleted, ...rest } = state.kits;
      return {
        ...state,
        length: state.length - 1,
        kits: {
          ...rest,
        },
      };
    }
    case '@KIT/UPDATE_PROPERTY': {
      const { [action.id]: updateable, ...rest } = state.kits;
      return {
        ...state,
        kits: {
          ...rest,
          [action.id]: kitReducer(updateable, action),
        },
      };
    }
    // these are straight map
    case '@MAP/RESIZE_WIDTH':
      return {
        ...state,
        width: action.value,
      };
    case '@MAP/RESIZE_HEIGHT':
      return {
        ...state,
        height: action.value,
      };
    case '@MAP/RENAME':
      return {
        ...state,
        name: action.value,
      };

    default:
      return state;
  }
}

export default mapReducer;
