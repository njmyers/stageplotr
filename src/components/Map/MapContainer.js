import * as React from 'react';
import { connect } from 'react-redux';
import Map from './Map';

const mapStateToProps = (state) => ({});

const mapDispatchToProps = (dispatch) => ({});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Map);
