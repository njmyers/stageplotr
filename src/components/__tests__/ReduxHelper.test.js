import * as React from 'react';
import { mount } from 'enzyme';
import ReduxHelper from './helpers/ReduxHelper';
import { createKit } from '../Map/map-actions';

const Dead = () => null;

describe('ReduxHelper component works as described', () => {
  test('we can dispatch from the mounted helper', () => {
    const wrapper = mount(
      <ReduxHelper>
        <Dead />
      </ReduxHelper>
    );

    const store = wrapper.state('store');
    store.dispatch(createKit('928d2ca0-94e4-11e8-82e4-61f5917152d6', 0));

    expect(store.getState().map.length).toBe(1);
    expect(store.getState()).toMatchSnapshot();
  });

  test('we can mount a new component with a new store', () => {
    const wrapper = mount(
      <ReduxHelper>
        <Dead />
      </ReduxHelper>
    );

    const store = wrapper.state('store');

    store.dispatch(createKit('928d2ca0-94e4-11e8-82e4-61f5917152d5', 0));
    expect(store.getState().map.length).toBe(1);

    store.dispatch(createKit('928d2ca0-94e4-11e8-82e4-61f5917152d6', 0));
    expect(store.getState().map.length).toBe(2);
    expect(store.getState()).toMatchSnapshot();
  });
});
