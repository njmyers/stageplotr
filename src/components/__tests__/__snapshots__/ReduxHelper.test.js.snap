// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduxHelper component works as described we can dispatch from the mounted helper 1`] = `
Object {
  "map": Object {
    "height": 100,
    "kits": Object {
      "928d2ca0-94e4-11e8-82e4-61f5917152d6": Object {
        "category": 0,
        "layer": 1,
        "orientation": 0,
        "positionX": 0,
        "positionY": 0,
        "rotation": 0,
        "scale": 100,
        "typeID": 0,
      },
    },
    "length": 1,
    "name": "StagePlot",
    "width": 100,
  },
}
`;

exports[`ReduxHelper component works as described we can mount a new component with a new store 1`] = `
Object {
  "map": Object {
    "height": 100,
    "kits": Object {
      "928d2ca0-94e4-11e8-82e4-61f5917152d5": Object {
        "category": 0,
        "layer": 1,
        "orientation": 0,
        "positionX": 0,
        "positionY": 0,
        "rotation": 0,
        "scale": 100,
        "typeID": 0,
      },
      "928d2ca0-94e4-11e8-82e4-61f5917152d6": Object {
        "category": 0,
        "layer": 1,
        "orientation": 0,
        "positionX": 0,
        "positionY": 0,
        "rotation": 0,
        "scale": 100,
        "typeID": 0,
      },
    },
    "length": 2,
    "name": "StagePlot",
    "width": 100,
  },
}
`;
